// osi/app/version-notification/index.js
import React, { Component } from 'react'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import Dialog from 'material-ui/Dialog'
import FlatButton from 'material-ui/FlatButton'
import * as userActions from '../../../../actions/user.actions'
import packageJSON from '../../../../../package.json'
import CHANGELOG from '../../../../../CHANGELOG.md'

console.log('CHANGELOG:', CHANGELOG)

class VersionNotificationDialog extends Component {
  constructor (props) {
    super(props)
    this.state = { isOpen: false }
  }

  componentDidMount () {
    const version = 'different version'
    // const version = packageJSON.version
    const lastVersion = this.props.user.properties.lastVersion

    // Check App version state
    console.info('has last version:', lastVersion)
    console.info('cur version:', version)
    console.log('version === lastVersion')
    console.log(version === lastVersion)
    console.log('version === lastVersion')

    if (version !== lastVersion) {
      // update user properties to reflect latestVersion
      // this.props.user.updateLatestVersion
      this.setState({isOpen: true})
    }
   }

  handleDismiss () {
    this.setState({isOpen: false})
  }

  render () {
    const actions = [
      <FlatButton
        label='Dismiss'
        primary
        onTouchTap={this.handleDismiss.bind(this)}
      />
    ]

    return (
      <div>
        <Dialog
          title='Version Update'
          actions={actions}
          modal
          open={this.state.isOpen}
        >
          Hello Eli!

          Latest version is 0.3.8

          <pre>
            {CHANGELOG}
          </pre>
        </Dialog>
      </div>
    )
  }
}

function mapStateToProps (state, ownProps) {
  return {
    user: state.user,
  }
}

function mapDispatchToProps (dispatch) {
  return {
    userActions: bindActionCreators(userActions, dispatch)
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(VersionNotificationDialog)
