// osi/app/version-notification/index.js
import React, { Component } from 'react'
import { connect } from 'react-redux'

import { bindActionCreators } from 'redux'
import semver from 'semver'

import Dialog from 'material-ui/Dialog'
import FlatButton from 'material-ui/FlatButton'
import RaisedButton from 'material-ui/RaisedButton'

import * as User from 'osi/user'

import * as userActions from '../../../../actions/user.actions'

import packageJSON from '../../../../../package.json'

import CHANGELOG from '../../../../../CHANGELOG.md'
console.log('CHANGELOG:', CHANGELOG)

class VersionNotificationDialog extends Component {
  constructor (props) {
    super(props)

    this.state = {
      isOpen: false
    }
  }

  componentWillUpdate (nextProps, nextState) {
    const { version } = packageJSON
    const lastVersion = nextProps && nextProps.userProps && nextProps.userProps.lastVersion
    // Check App version state
    if (!nextProps.userProps || !nextProps.userProps.lastVersion) {
      console.log('no props, no last version')
      // Show welcome version
      if (!nextProps.user || !nextProps.user.showVersionNotification) {
        console.log('no notification show')
        this.props.userActions.showVersionNotification()
      } else if (nextProps.user && nextProps.user.showVersionNotification && !nextState.isOpen) {
        this.setState({isOpen: true})
      }
    } else if (nextProps.userProps.lastVersion) {
      console.info('has last version:', lastVersion)
      console.info('cur version:', version)
      if (semver.gt(version, lastVersion)) {
        console.log('LAST VERSION IS OLDER THAN NEW VERSION')
      } else {
        this.setState({isOpen: true})

        console.log('LAST VERSION IS NOT OLDER THAN NEW VERSION')
      }
    }
  }

  handleOpen () {
    this.setState({open: true})
  }

  handleDismiss () {
    const { uid } = this.props.user.profile
    const { version } = packageJSON
    const { lastVersion } = this.props.userProps

    if (!lastVersion) {
      console.log('not last version')
      this.props.userActions.hideVersionNotification(uid, version)
    } else {
      console.info('has last version:', lastVersion)
      console.info('cur version:', version)
    }
    // this.props.userActions.hideVersionNotification(uid)
    // this.setState({open: false})
  }

  render () {
    const actions = [
      <FlatButton
        label='Dismiss'
        primary
        onTouchTap={this.handleDismiss.bind(this)}
      />
    ]

    return (
      this.props.user.showVersionNotification ? (
        <Dialog
          title='Version Update'
          actions={actions}
          modal
          open={true || this.state.isOpen}
        >
          Hello Eli!

          Latest version is 0.3.8

          <pre>
            {CHANGELOG}
          </pre>
        </Dialog>
      ) : null
    )
  }
}

function mapStateToProps (state, ownProps) {
  const { user } = state
  return {
    user,
    userProps: user && user.properties
  }
}

function mapDispatchToProps (dispatch) {
  return {
    userActions: bindActionCreators(userActions, dispatch)
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(VersionNotificationDialog)
