// osi/components/header/header.react.js
import React from 'react'
import { connect } from 'react-redux'
import { Link, browserHistory } from 'react-router'
import { Navbar, Nav, NavItem, Modal, Button, ControlLabel, FormControl, HelpBlock, FormGroup } from 'react-bootstrap'
import { LinkContainer } from 'react-router-bootstrap'
import firebase from '../../../../utils/database.react.js'
import FileUploader from 'react-firebase-file-uploader'
import { Icon } from 'osi/bits'
import { bindActionCreators } from 'redux'
import * as userActions from '../../../../actions/user.actions'


import './header.css'

const { localStorage, alert } = window

class Header extends React.Component {
  constructor (props, context) {
    super(props, context)

    this.state = {showHelp: null, showShare: false, userName: null, image: null}

    this.closeHelp = this.closeHelp.bind(this)
    this.openHelp = this.openHelp.bind(this)
    this.isShowHelp = this.isShowHelp.bind(this)
    this.closeShare = this.closeShare.bind(this)
    this.openShare = this.openShare.bind(this)
    this.share = this.share.bind(this)
    this.saveUserName = this.saveUserName.bind(this)
    this.handleUploadStart = this.handleUploadStart.bind(this)
    this.handleUploadError = this.handleUploadError.bind(this)
    this.handleUploadSuccess = this.handleUploadSuccess.bind(this)
    this.handleProgress = this.handleProgress.bind(this)
  }

  componentWillMount () {

  }

  openHelp (event) {
    this.setState({ showHelp: true })
  }

  closeHelp (event) {
    this.setState({ showHelp: false })
  }

  openShare (event) {
    // this.setState({ showShare: true })
  }

  closeShare (event) {
    this.setState({ showShare: false, image: null, userName: null })
  }

  isShowHelp () {
    const allowedHosts = ['orymeli:3000', 'coindash.herokuapp.com']
    const isAllowedHost = !(allowedHosts.indexOf(window.location.host) === -1)
    return this.props.extension.version === '0.0.0' && (this.state.showHelp === null || this.state.showHelp) && !isAllowedHost
  }

  saveUserName (event) {
    this.setState({userName: event.target.value})
  }

  share (event) {
    if (this.state.image === null || this.state.userName === null) {} else {
      if (this.props.chartLoaded === true) {
        var id = getDataFromLocalStorage()
        var key = firebase.database().ref().push().key
        var ref = firebase.database().ref('users')

        var userData = {
          id: key,
          performanceData: this.props.performanceData,
          balances: this.props.balances,
          exchanges: this.props.exchanges,
          chartData: this.props.chartData,
          shortDelta: this.props.shortDelta,
          longDelta: this.props.longDelta,
          copiers: 0,
          userName: this.state.userName,
          image: this.state.image
        }
        if (id === null || id === 'default' || id === '') {
          userData.id = key
          ref.child(key).set(userData)
          saveDataToLocalStorage(key)
          console.log('saved in ls + fb')
        } else {
          userData.id = id
          ref.child(id).set(userData)
          console.log('saved in fb')
        }
      } else {
        console.log('chart not loaded')
        alert('Please load chart first')
      }
      this.setState({uploaded: false})
      this.closeShare()
    }
  }

  getUserNameValidationState () {
    if (this.state.userName != null) {
      return 'success'
    } else {
      return 'error'
    }
  }

  getImageValidationState () {
    if (this.state.uploaded) {
      return 'success'
    } else {
      return 'error'
    }
  }

  handleUploadStart (file) {
    this.setState({isUploading: true, progress: 0})
  }

  handleProgress (progress) { return this.setState({progress}) }

  handleUploadError (error) {
    this.setState({isUploading: false, progress: 0})
    console.error(error)
  }

  handleUploadSuccess (filename) {
    this.setState({avatar: filename, progress: 100, isUploading: false, uploaded: true})
    firebase.storage().ref('images')
      .child(filename)
      .getDownloadURL()
      .then(url => { this.setState({ image: url }) })
  }

  handleLogout (event) {
    event.preventDefault()
    this.props.userActions.logout()

    browserHistory.push('/')

  }

  render () {
    return (
      <header>
        <Navbar className='cd-navbar'>
          <Navbar.Header>
            <Navbar.Brand>
              <Link to='/'>
                <Icon />
              </Link>
            </Navbar.Brand>
          </Navbar.Header>
          <Nav pullRight>
            {false && (
              <LinkContainer className='disabled' onClick={this.openShare} to='share'>
                <NavItem eventKey={1}>Share</NavItem>
              </LinkContainer>
            )}
            {false && (
              <LinkContainer onClick={this.openHelp} to='help'>
                <NavItem eventKey={1}>Help</NavItem>
              </LinkContainer>
            )}
            <LinkContainer onClick={this.handleLogout.bind(this)} to='logout'>
              <NavItem>Log out</NavItem>
            </LinkContainer>
          </Nav>
        </Navbar>

        <Modal show={this.state.showShare} onHide={this.closeShare}>
          <Modal.Header closeButton>
            <Modal.Title>Share Your Profile</Modal.Title>
          </Modal.Header>
          <Modal.Body>
            <form >
              <FormGroup controlId='formToken' validationState={this.getUserNameValidationState()}>
                <ControlLabel>UserName</ControlLabel>
                <FormControl type='text' placeholder='Enter userName' onChange={this.saveUserName} />
                <FormControl.Feedback />
              </FormGroup>
              <FormGroup controlId='formToken' validationState={this.getImageValidationState()}>
                <ControlLabel>Profile Image</ControlLabel>
                <FileUploader
                  accept='image/*'
                  name='avatar'
                  randomizeFilename
                  storageRef={firebase.storage().ref('images')}
                  onUploadStart={this.handleUploadStart}
                  onUploadError={this.handleUploadError}
                  onUploadSuccess={this.handleUploadSuccess}
                  onProgress={this.handleProgress}
            />
                <HelpBlock>Please Upload Image</HelpBlock>
              </FormGroup>
            </form>
          </Modal.Body>
          <Modal.Footer>
            <Button onClick={this.share} >Share</Button>
            <Button onClick={this.closeShare} >Cancel</Button>
          </Modal.Footer>
        </Modal>
      </header>
    )
  }
}

function mapStateToProps (state, ownProps) {
  return {
    balances: state.balances,
    chartData: state.charts.chartData,
    performanceData: state.charts.preformanceData,
    front: state.coins.front,
    exchanges: state.exchanges,
    statusText: state.charts.statusText,
    chartLoaded: state.charts.chartLoaded,
    shortDelta: state.charts.shortDelta,
    longDelta: state.charts.longDelta
  }
}

function saveDataToLocalStorage (obj) {
  localStorage.setItem('id', JSON.stringify(obj))
}

function getDataFromLocalStorage (obj) {
  if (localStorage.getItem('id') != null) { return localStorage.getItem('id').substring(1, localStorage.getItem('id').length - 1) } else { return '' }
}


function mapDispatchToProps (dispatch) {
  return {
    userActions: bindActionCreators(userActions, dispatch),
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(Header)
