// osi/components/portman/portfolio-management/home/index.js
import React, { Component } from 'react'
import { connect } from 'react-redux'

import { bindActionCreators } from 'redux'

import * as userActions from '../../../../../../actions/user.actions'
import * as portfolioActions from '../../../../../../actions/portfolio.actions'

import RaisedButton from 'material-ui/RaisedButton'

import FlatButton from 'material-ui/FlatButton'
import Dialog from 'material-ui/Dialog'
import TextField from 'material-ui/TextField'

import PortfoliosList from './portfolios-list'

class CreatePortfolioDialog extends Component {
  constructor (props) {
    super(props)

    this.state = {
      allowSubmit: false
    }
  }

  onSubmit (e) {
    const { uid } = this.props.user.profile
    const name = this.refs.name.getValue()
    const description = this.refs.description.getValue()

    const { onCreate = () => {} } = this.props
    e.preventDefault()

    this.props.portfolioActions.newPortfolio({name, description, uid})
      .then(pmResult => {
        onCreate(pmResult)
      })
  }
  valuesUpdated (e) {
    const name = this.refs.name.getValue()
    const { allowSubmit } = this.state
    if ((allowSubmit && !name) || (!allowSubmit && name)) {
      this.setState({allowSubmit: !!name})
    }
  }
  render () {
    const { allowSubmit } = this.state
    return (
      <Dialog
        title='Create Portfolio'
        onRequestClose={e => this.props.onCancel(e)}
        actions={([
          <FlatButton
            label='Cancel'
            primary
            onTouchTap={e => this.props.onCancel(e)}
          />,
          <FlatButton
            label='Submit'
            primary
            disabled={!allowSubmit}
            onTouchTap={e => this.onSubmit(e)}
          />
        ])}
        open={this.props.isOpen}
      >
        <form onSubmit={e => this.onSubmit(e)}>
          <div>
            <TextField
              ref='name'
              floatingLabelText='Portfolio name'
              onChange={e => this.valuesUpdated()}
            />
          </div>
          <div>
            <TextField
              ref='description'
              multiLine
              floatingLabelText='Description'
              hintText='Give your portfolio a description'
            />
          </div>
        </form>
      </Dialog>
    )
  }
}

const ConnectedCreatePortfolioDialog = connect(mapStateToProps, mapDispatchToProps)(CreatePortfolioDialog)

class Home extends Component {
  constructor (props) {
    super(props)

    this.state = {
      createPortfolioOpen: false
    }
  }

  createPortfolio () {
    // console.log('lets create a portfolio')

    this.setState({createPortfolioOpen: true})

    // this.props.portfolioActions.newPortfolio({
    //   name: 'Elis first portfolio',
    //   description: 'My first!',
    //   uid
    // })
    //   .then(pmResult => {
        // console.info('portfolio creation result:', pmResult)
    //   })
  }
  clearPortfolios () {
    // this.props.portfolioActions.clearPortfolios()
  }
  onPortfolioCreated (dat) {
    // console.info('what is dat?', dat)
    this.props.portfolioActions.loadUserPortfolios(this.props.user.profile.uid)
      .then(res => {
        // console.log('what is res?', res)

        this.setState({createPortfolioOpen: false})
      })
  }
  onCellClicked (row, cell, event) {
    const { portfolio } = this.props
    const { userPortfolios = [] } = portfolio

    // console.log('row/cell clicked:', row, cell)
    console.log('row, cell, event', row, cell, event)
    // console.log('more', more)
    this.props.portfolioActions.setCurrentPortfolio(userPortfolios[row])
  }
  render () {
    const { userPortfolios } = this.props
    // console.info('user portfolios:', userPortfolios, 'portfolio', portfolio)
    // console.info('portfoliosList:', PortfoliosList)
    return (
      <div className='home'>
        <h2>My Portfolios</h2>

        <RaisedButton label='Create new Portfolio' onClick={e => this.createPortfolio()} />
        <RaisedButton label='Clear Portfolios' onClick={e => this.clearPortfolios()} />

        <hr />

        <PortfoliosList
          portfolios={userPortfolios}
          onCellClick={(...args) => this.onCellClicked(...args)}
          onCreatePortfolio={e => this.createPortfolio()} />

        <ConnectedCreatePortfolioDialog
          onCancel={e => this.setState({createPortfolioOpen: false})}
          onCreate={pf => this.onPortfolioCreated(pf)}
          isOpen={this.state.createPortfolioOpen} />
      </div>
    )
  }
}

const getUserPortfolios = ({portfolio}) => {
  if (!portfolio || !portfolio.userPortfolios) return []

  return portfolio.userPortfolios.map(pid =>
    portfolio.portfolios.find(e => e.portfolio.pid === pid)
  )
}

function mapStateToProps (state, ownProps) {
  return {
    user: state.user,
    portfolios: state.portfolio.portfolios,
    userPortfolios: getUserPortfolios(state),
    portfolio: state.portfolio
  }
}

function mapDispatchToProps (dispatch) {
  return {
    userActions: bindActionCreators(userActions, dispatch),
    portfolioActions: bindActionCreators(portfolioActions, dispatch)
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(Home)
