// osi/components/sidenav2/index.js
import React, { Component } from 'react'
import { Link } from 'react-router'
import { connect } from 'react-redux'
import SidenavItem from './sidenav-item.react'
import Paper from 'material-ui/Paper'
import MenuItem from 'material-ui/MenuItem'
import RemoveRedEye from 'material-ui/svg-icons/image/remove-red-eye'
import PersonAdd from 'material-ui/svg-icons/social/person-add'
import ContentLink from 'material-ui/svg-icons/content/link'
import ContentCopy from 'material-ui/svg-icons/content/content-copy'
import Download from 'material-ui/svg-icons/file/file-download'
import Delete from 'material-ui/svg-icons/action/delete'
import FontIcon from 'material-ui/FontIcon'

import Menu from 'react-md/lib/Menus/Menu'
import List from 'react-md/lib/Lists/List'
import ListItem from 'react-md/lib/Lists/ListItem'
import Divider from 'react-md/lib/Dividers'

import ActionHome from 'material-ui/svg-icons/action/home'
import ActionCardTravel from 'material-ui/svg-icons/action/card-travel'
import ActionDoneAll from 'material-ui/svg-icons/action/done-all'

import './sidenav.css'

class SideNavigation extends Component {
  defaultProps () {
    return { items: [
            {title: 'dashboard', icon: 'tachometer'}
    ] }
  }

  render () {
    const disabled = true
    const menuItems = [
            {title: 'Dashboard', icon: <ActionHome />, route: '/dashboard'},
            {title: 'Portfolio', icon: <ActionCardTravel />, route: '/portfolios'},
            // {title: 'My Accounts', icon: 'cogs', route: '/dashboard/accounts', style: {paddingTop: '10px'}},
            // {title: 'My Tokens', icon: 'cubes', route: '/dashboard/tokens', style: {paddingTop: '10px'}},
            {title: 'Copy ICO', icon: <ActionDoneAll />, route: '/copyCrypto', style: { paddingTop: '11px' }, disabled},
            {title: 'about', icon: 'university', route: '/about', style: { paddingTop: '8px' }, disabled}
            /* {title: 'about', icon: 'building-o', route: '/accounts', style: {paddingTop: '10px'}},
            {title: 'links', icon: 'link', route: '/accounts', style: {paddingTop: '11px'}} */
    ]

    const adminItems = [
            { title: 'admin', icon: 'database', route: '/admin' },
            { title: 'users', icon: 'users', route: '/admin/users' },
            { title: 'Portfolios', icon: 'newspaper-o', route: '/admin/portfolios' }
            // {title: 'Portfolio', icon: 'newspaper-o', route: '/portfolio', style: {paddingTop: '10px'}},
            // {title: 'My Accounts', icon: 'cogs', route: '/dashboard/accounts', style: {paddingTop: '10px'}},
            // {title: 'My Tokens', icon: 'cubes', route: '/dashboard/tokens', style: {paddingTop: '10px'}},
            // {title: 'Copy ICO', icon: 'newspaper-o', route: '/copyCrypto', style: { paddingTop: '11px' }, disabled},
            // {title: 'about', icon: 'university', route: '/about', style: { paddingTop: '8px' }, disabled}
            /* {title: 'about', icon: 'building-o', route: '/accounts', style: {paddingTop: '10px'}},
            {title: 'links', icon: 'link', route: '/accounts', style: {paddingTop: '11px'}} */
    ]
    return (
      <Paper zDepth={1} className='sidenav' style={{borderRadius: 0}}>

        <List className='main'>
          {menuItems.map((m, i) =>
            <ListItem activeClassName='active' to={m.route} key={i} primaryText={m.title} leftIcon={m.icon} component={Link} />
          )}
          <ListItem primaryText='Share' leftIcon={<PersonAdd />} />
          <ListItem primaryText='Get links' leftIcon={<ContentLink />} />
          <Divider />
          <ListItem primaryText='Make a copy' leftIcon={<ContentCopy />} />
          <ListItem primaryText='Download' leftIcon={<Download />} />
          <Divider />
          <ListItem primaryText='Remove' leftIcon={<Delete />} />
        </List>

        <ul>
          {menuItems.map((item, index) => item.title)}
        </ul>
        <div className='spacer' />
        {this.props.isAdmin && (
          <div className='admin'>
            <ul>
              {adminItems.map((item, index) => item.title)}
            </ul>
          </div>
        )}
      </Paper>
    )
  }
}

function mapStateToProps (state, ownProps) {
  return {
    user: state.user,
    isAdmin: state.admin.active,
    portfolios: state.portfolio
  }
}

function mapDispatchToProps (dispatch) {
  return {
  }
}
export default connect(mapStateToProps, mapDispatchToProps)(SideNavigation)
