// osi/components/sidenav/sidenav.react.js
import React, { Component } from 'react'
import { connect } from 'react-redux'
import SidenavItem from './sidenav-item.react'
import './sidenav.css'

class SideNavigation extends Component {
  defaultProps () {
    return { items: [
            {title: 'dashboard', icon: 'tachometer'}
    ] }
  }

  render () {
    const disabled = true
    const menuItems = [
            {title: 'dashboard', icon: 'tachometer', route: '/dashboard', style: {paddingTop: '6px'}},
            {title: 'Portfolio', icon: 'newspaper-o', route: '/portfolios', style: {paddingTop: '10px'}},
            // {title: 'My Accounts', icon: 'cogs', route: '/dashboard/accounts', style: {paddingTop: '10px'}},
            // {title: 'My Tokens', icon: 'cubes', route: '/dashboard/tokens', style: {paddingTop: '10px'}},
            {title: 'Copy ICO', icon: 'newspaper-o', route: '/copyCrypto', style: { paddingTop: '11px' }, disabled},
            {title: 'about', icon: 'university', route: '/about', style: { paddingTop: '8px' }, disabled}
            /* {title: 'about', icon: 'building-o', route: '/accounts', style: {paddingTop: '10px'}},
            {title: 'links', icon: 'link', route: '/accounts', style: {paddingTop: '11px'}} */
    ]

    const adminItems = [
            { title: 'admin', icon: 'database', route: '/admin' },
            { title: 'users', icon: 'users', route: '/admin/users' },
            { title: 'Portfolios', icon: 'newspaper-o', route: '/admin/portfolios' }
            // {title: 'Portfolio', icon: 'newspaper-o', route: '/portfolio', style: {paddingTop: '10px'}},
            // {title: 'My Accounts', icon: 'cogs', route: '/dashboard/accounts', style: {paddingTop: '10px'}},
            // {title: 'My Tokens', icon: 'cubes', route: '/dashboard/tokens', style: {paddingTop: '10px'}},
            // {title: 'Copy ICO', icon: 'newspaper-o', route: '/copyCrypto', style: { paddingTop: '11px' }, disabled},
            // {title: 'about', icon: 'university', route: '/about', style: { paddingTop: '8px' }, disabled}
            /* {title: 'about', icon: 'building-o', route: '/accounts', style: {paddingTop: '10px'}},
            {title: 'links', icon: 'link', route: '/accounts', style: {paddingTop: '11px'}} */
    ]
    return (
      <div className='sidenav'>
        <ul>
          {menuItems.map((item, index) => <SidenavItem key={index} item={item} />)}
        </ul>
        <div className='spacer' />
        {this.props.isAdmin && (
          <div className='admin'>
            <ul>
              {adminItems.map((item, index) => <SidenavItem key={index} item={item} />)}
            </ul>
          </div>
        )}
      </div>
    )
  }
}

function mapStateToProps (state, ownProps) {
  return {
    user: state.user,
    isAdmin: state.admin.active,
    portfolios: state.portfolio
  }
}

function mapDispatchToProps (dispatch) {
  return {
  }
}
export default connect(mapStateToProps, mapDispatchToProps)(SideNavigation)
