// osi/components/portman/portfolio-management/dashboard/index.js
import React, { Component } from 'react'
import { connect } from 'react-redux'

import Paper from 'react-md/lib/Papers'
import CardText from 'react-md/lib/Cards/CardText'
import Button from 'material-ui/RaisedButton'

import './dashboard.css'

class Dashboard extends Component {
  constructor (props) {
    super(props)

    this.state = {}
  }

  render () {
    const { user = {} } = this.props
    return (
      <div className='dashboard'>
        <h1>Welcome, {user.profile.name}!</h1>
        <WelcomeBlock />
      </div>
    )
  }
}

const WelcomeBlock = (props) => {
  const imgSrc = '/images/cryptos.png'
  return (
    <Paper className='welcome' zDepth={2}>
      <div className='pres'>
        <img src={imgSrc} role='presentation' />
      </div>
      <div className='details'>
        <h3>Get started with creating your first portfolio</h3>
        <p>
         Portfolios are your way to track the earnings and performance of your investments.
       </p>
        <p>
         Coindash portfolios support ETH and other ERC20 tokens. More to come soon...
       </p>
       <div className='actions'>
         <Button label='Create' primary />
       </div>
      </div>
    </Paper>
  )
}

/*

 <div className='detail'>
   <h3>Get started with creating your first portfolio</h3>
   <p>
     Portfolios are your way to track the earnings and performance of your investments.
   </p>
   <p>
     Coindash portfolios support ETH and other ERC20 tokens. More to come soon...
   </p>
 </div>
 */
function mapStateToProps (state, ownProps) {
  const { portfolios, userPortfolios: userKeys } = state.portfolio
  const userPortoflios = portfolios.filter(pf => userKeys.indexOf(pf) >= 0)

  return {
    user: state.user,
    userPortoflios,
    portfolios
  }
}

function mapDispatchToProps (dispatch) {
  return {
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(Dashboard)
