// cd-admin/admin/index.js
import React, { Component } from 'react'

import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'

import * as userActions from '../../../actions/user.actions'

import * as adminActions from '../../../actions/admin.actions'

import getMuiTheme from 'material-ui/styles/getMuiTheme'
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider'
import lightBaseTheme from 'material-ui/styles/baseThemes/lightBaseTheme'

import Paper from 'material-ui/Paper'

import './admin.css'

class Wrap extends Component {
  render () {
    return (
      <MuiThemeProvider muiTheme={getMuiTheme(lightBaseTheme)}>
        <Paper className='admin-wrap' zDepth={1}>
          <h1>Welcome to the admin world.</h1>
          {this.props.children}
        </Paper>
      </MuiThemeProvider>

    )
  }
}

export const AdminWrap = connect(mapStateToProps, mapDispatchToProps)(Wrap)

class Home extends Component {
  render () {
    return (
      <div>
        <h3>Home</h3>
      </div>
    )
  }
}

export const AdminHome = connect(mapStateToProps, mapDispatchToProps)(Home)

function mapStateToProps (state, ownProps) {
  return {
    user: state.user,
    isAdmin: state.admin.active
  }
}

function mapDispatchToProps (dispatch) {
  return {
    userActions: bindActionCreators(userActions, dispatch),
    adminActions: bindActionCreators(adminActions, dispatch)
  }
}
