// osi/themes/coindashBaseTheme.js
// import React from 'react'
// import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider'
// import AppBar from 'material-ui/AppBar'
import lightBaseTheme from 'material-ui/styles/baseThemes/lightBaseTheme'
import { defaultsDeep } from 'lodash'

import {
  lightBlue900,
  lightBlue700,
  lightBlue500,
  lightGreen900,
  lightGreen700,
  lightGreen500,
  darkBlack,
  white,
  grey300
} from 'material-ui/styles/colors'

// This replaces the textColor value on the palette
// and then update the keys for each component that depends on it.
// More on Colors: http://www.material-ui.com/#/customization/colors
const coindashBaseTheme = defaultsDeep({
  spacing: {
    iconSize: 24
  },
  palette: {
    primary1Color: lightBlue500,
    primary2Color: lightBlue700,
    primary3Color: lightBlue900,
    accent1Color: lightGreen500,
    accent2Color: lightGreen700,
    accent3Color: lightGreen900,
    textColor: darkBlack,
    alternateTextColor: white,
    canvasColor: white,
    borderColor: grey300
  }
}, lightBaseTheme)

export default coindashBaseTheme
