// osi/themes/coindashDarkTheme.js
// import React from 'react'
// import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider'
// import AppBar from 'material-ui/AppBar'
import baseTheme from './coindashBaseTheme'
import {fade} from 'material-ui/utils/colorManipulator'
import { defaultsDeep } from 'lodash'
import {
  cyan500,
  indigo900,
  orange200,
  deepOrange300,

  lightBlue900,
  lightBlue700,
  lightBlue500,
  lightBlue400,

  lightGreen900,
  lightGreen700,
  lightGreen500,
  lightGreen400,

  cyan700,
  grey400,
  pinkA200,
  grey100,
  grey500,
  darkBlack,
  white,
  grey300,
  fullBlack,
  blueGrey900,
  blueGrey800,
  blueGrey50
} from 'material-ui/styles/colors'

// This replaces the textColor value on the palette
// and then update the keys for each component that depends on it.
// More on Colors: http://www.material-ui.com/#/customization/colors
const coindashDarkTheme = defaultsDeep({
  palette: {
    textColor: white,
    alternateTextColor: blueGrey50,
    canvasColor: blueGrey900
  },
  raisedButton: {
    color: blueGrey800,
    disabledColor: blueGrey800
  }
}, baseTheme)

export default coindashDarkTheme
