// osi/themes/coindashBaseTheme.js
import lightBaseTheme from 'material-ui/styles/baseThemes/lightBaseTheme'
import { defaultsDeep } from 'lodash'

const primary1Color = '#4845D3'

import {
  lightBlue900,
  lightBlue700,
  lightBlue500,
  lightGreen900,
  lightGreen700,
  lightGreen500,
  darkBlack,
  white,
  grey300
} from 'material-ui/styles/colors'

// This replaces the textColor value on the palette
// and then update the keys for each component that depends on it.
// More on Colors: http://www.material-ui.com/#/customization/colors
const coindashBaseTheme = defaultsDeep({
  spacing: {
    iconSize: 24
  },
  palette: {
    primary1Color: primary1Color,
    primary2Color: lightBlue700,
    primary3Color: lightBlue900,
    accent1Color: lightGreen500,
    accent2Color: lightGreen700,
    accent3Color: lightGreen900,
    textColor: darkBlack,
    alternateTextColor: white,
    canvasColor: white,
    borderColor: grey300
  },
  tabs: {
    backgroundColor: 'transparent',
    textColor: darkBlack,
    selected: primary1Color,
    selectedTextColor: primary1Color
  },
  inkBar: {
    backgroundColor: primary1Color
  }
}, lightBaseTheme)

export default coindashBaseTheme
