// osi/themes/coindashBaseTheme.js
// import React from 'react'
// import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider'
// import AppBar from 'material-ui/AppBar'
import {fade} from 'material-ui/utils/colorManipulator'
import lightBaseTheme from 'material-ui/styles/baseThemes/lightBaseTheme'
import { defaultsDeep } from 'lodash'

import {
  cyan500,
  indigo900,
  orange200,
  deepOrange300,

  lightBlue900,
  lightBlue700,
  lightBlue500,
  lightBlue400,

  lightGreen900,
  lightGreen700,
  lightGreen500,
  lightGreen400,

  cyan700,
  grey400,
  pinkA200,
  grey100,
  grey500,
  darkBlack,
  white,
  grey300,
  fullBlack
} from 'material-ui/styles/colors'

// This replaces the textColor value on the palette
// and then update the keys for each component that depends on it.
// More on Colors: http://www.material-ui.com/#/customization/colors
const coindashBaseTheme = defaultsDeep({
  spacing: {
    iconSize: 24,

    desktopGutter: 24,
    desktopGutterMore: 32,
    desktopGutterLess: 16,
    desktopGutterMini: 8,
    desktopKeylineIncrement: 64,
    desktopDropDownMenuItemHeight: 32,
    desktopDropDownMenuFontSize: 15,
    desktopDrawerMenuItemHeight: 48,
    desktopSubheaderHeight: 48,
    desktopToolbarHeight: 56
  },
  appBar: {
    height: 50,
    color: white,
    textColor: darkBlack
  },
  palette: {
    primary1Color: lightBlue500,
    primary2Color: lightBlue700,
    primary3Color: lightBlue900,
    accent1Color: lightGreen500,
    accent2Color: lightGreen700,
    accent3Color: lightGreen900,
    textColor: darkBlack,
    alternateTextColor: white,
    canvasColor: white,
    borderColor: grey300,
    disabledColor: fade(darkBlack, 0.3),
    pickerHeaderColor: cyan500,
    clockCircleColor: fade(darkBlack, 0.07),
    shadowColor: fullBlack
  }
}, lightBaseTheme)

export default coindashBaseTheme
