// osi/components/portman/portfolio-management/dashboard/benchmark/index.js
import React, { Component } from 'react'
import _ from 'lodash'

import Paper from 'react-md/lib/Papers'
import { Tabs, Tab } from 'material-ui/Tabs'

import ActionTrendingUp from 'material-ui/svg-icons/action/trending-up'
import EditorPieChart from 'material-ui/svg-icons/editor/pie-chart'

import PerformanceChart from 'osi/components/portman/parts/charts/performance'
import AllocationChart from 'osi/components/portman/parts/charts/allocation'
import colors from 'styleguide/colors'

const TabIcons = {
  performance: props => <ActionTrendingUp {...props} />,
  allocation: props => <EditorPieChart {...props} />

}

const TabName = ({icon, ...props}) => {
  const TabIcon = TabIcons[icon]

  return (
    <div className='tab-name'>
      <TabIcon style={{width: 15}} />
      <span>{props.children}</span>
    </div>
  )
}

class Benchmark extends Component {
  constructor (props) {
    super(props)

    this.state = {
      activeTab: 'b'
    }
  }
  handleChange (...args) {
    console.info('changed', args)
  }

  render () {
    const { balances, calculations, onboarding } = this.props
    const calcs = _.cloneDeep(calculations)
    const bals = _.cloneDeep(balances)

    return (
      <Paper className='benchmark-card'>
        <h4>Benchmark</h4>

        <Tabs
          className='tabs'
          value={this.props.activeTab}
          onChange={this.props.handleChange}
        >
          <Tab label={(<TabName icon='performance'>Performance</TabName>)} value='a' className='performance'>
            {!onboarding && calcs && (
              <PerformanceChart calculations={calcs} />
            )}
            {onboarding && <OnBoardingMessage />}

          </Tab>
          <Tab  label={<TabName icon='allocation'>Allocation</TabName>} value='b' className='allocation'>
            {!onboarding && balances && !balances.fetching && balances.data && balances.data.length && (
              <AllocationChart balances={bals.data} colors={colors[balances.data.length]} />
            )}
            {onboarding && <OnBoardingMessage />}
          </Tab>
        </Tabs>
      </Paper>
    )
  }
}

const OnBoardingMessage = props => (
  <div className='onboarding-message'>
    <Paper>
      <p>
        Create your first crypto portfolio and get insights into their
        perfomances, risks, and allocations.
      </p>
    </Paper>
  </div>
)

export default Benchmark
