// osi/analytics/index.js
// import config from '../../config'
import KeenTracking from 'keen-tracking'

import packageJson from '../../../../package.json'
// Configure a client instance
var client = new KeenTracking({
  projectId: '592ee49c3d5e1597cc43a840',
  writeKey: 'E1716BECBDA3D2CA47D59C90CCCDED7322A91ED87210D35125F54E14B453B78604BB401846AABA771D03ED72B2C2A4BC69CCF463CA75421CD3167B2FC93DD1D72021964CEAE239D1E2C33C67E32D99E40D74628FE79CF5E74B4F146675973CBF'
})
let loggingDisabled = false

client.addEvent('app-load', {
  ip_address: `${'$'}{keen.ip}`,
  keen: {
    addons: [
      {
        name: 'keen:ip_to_geo',
        input: {
          ip: 'ip_address'
        },
        output: 'ip_geo_info'
      }
    ]
  }
})

let myglobs = (eventCollection) => {
  let props = {
    version: packageJson.version,
    landing: window.location.toString(),
    hostname: window.location.host,
    env: process.env.NODE_ENV
  }

  return props
}
client.setGlobalProperties(myglobs)

client.recordEvent('app-load', {
  loadTime: Date.now() - window.startLoad
})

export const listener = location => {
  if (!loggingDisabled) {
    client.recordEvent('pageviews', {
      page: location.pathname
    })
  }
}

export const recordEvent = (name, opts, cb) => {
  if (!loggingDisabled) {
    client.recordEvent(name, opts, cb)
  }
}

export const recordEvents = (events, cb) => {
  if (!loggingDisabled) {
    client.recordEvents(events, cb)
  }
}

export const disable = () => {
  loggingDisabled = true
}

export const enable = () => {
  loggingDisabled = false
}
