// osi/user/coupons.js
import firebase from '../../../utils/database.react.js'
import { getCurrentUser } from 'osi/auth'
import { usersPath } from './'

export const metaPath = 'meta'
export const couponsPath = `${metaPath}/coupons`
export const presalePath = `${metaPath}/presale`

const db = firebase.database()

const getUserRef = (uid) => db.ref(`${couponsPath}/${uid}`)

export const isCouponValid = (coupon) => {
  console.info('validate coupon:', coupon)
  return new Promise((resolve, reject) => db.ref(couponsPath).child(coupon)
    .once('value')
    .then(snp => {
      const value = snp.val()
      console.log('value of coupon:', value)
      resolve(!!value && !value.used)
    })
    .catch(err => {
      console.log('error validating coupon:', err)
      if (err.code === 'PERMISSION_DENIED') {
        reject(err)
      }
    })
  )
}

export const addCoupon = (coupon) => {
  let cRef = db.ref(couponsPath).child(coupon)

  return cRef.set({
    used: false
  })

}

export const useCoupon = (coupon, uid) => {
  let updates = {}
  updates[`${couponsPath}/${coupon}/used`] = true
  updates[`${usersPath}/${uid}/coupon`] = {
    couponCode: coupon,
    usedOn: firebase.database.ServerValue.TIMESTAMP
  }

  console.log('useCoupon:', coupon, uid)
  console.info('updates', updates)

  return db.ref().update(updates)
}

export const getCouponValue = () => {
  let aRef = db.ref(presalePath).child('code')

  return new Promise((resolve, reject) => {
    aRef.once('value')
    .then(snp => {
      const value = snp.val()
      console.info('getCouponValue value:', value)
      resolve(value)
    })
    .catch(err => {
      console.info('error getting coupon value:', err)
      reject(err)
    })
  })
}

export const userHasCoupon = (uid) => {
  return new Promise((resolve, reject) => db.ref(usersPath).child(uid).child('coupon')
    .once('value')
    .then(snp => {
      const value = snp.val()
      console.log('value of coupon:', value)
      resolve(!!value && !value.used)
    })
    .catch(err => {
      console.log('error validating coupon:', err)
      if (err.code === 'PERMISSION_DENIED') {
        reject(err)
      }
    })
  ).then(hasCoupon => hasCoupon && db.ref(presalePath).child('code')
    .once('value')
    .then(snp => snp.val()))
}

getCurrentUser().then(user => {
  // console.log('our user is:', user)
  // console.info('our db is:', db)

  // Check if user is admin
  // isAdmin(user.uid).then(d => {
    // console.log('isAdmin returned', d)
  // })
})

export const createNewUser = ({uid, ...rest}) => {
  const userRef = getUserRef(uid)

  // make sure the `uid` doesn't exist...
  let user = {
    createdOn: firebase.database.ServerValue.TIMESTAMP,
    ...rest
  }

  return userRef.set(user)
    .catch(err => {
      console.error('user add failed:', err)
    })
}

export const isAdmin = uid => {
  // console.info('checking if admin for', uid)
  let adminRef = db.ref(`admin/${couponsPath}/${uid}`)

  return new Promise((resolve, reject) => {
    adminRef.once('value')
    .then(snp => {
      const val = snp.val()
      resolve(val !== null && ((val.gmode && 'GMODE') || true))
    })
    .catch(err => {
      // console.log('admin/users/', uid, 'returned error:', err)
      if (err.code === 'PERMISSION_DENIED') {
        resolve(false)
      } else {
        reject(err)
      }
    })
  })
}

export const getUser = uid => new Promise((resolve, reject) => {
  let prom = db.ref(couponsPath).child(uid).once('value')

  prom
    .then(snp => snp.val())
    .then(val => {
      return val
    })
    .then(resolve)
})

export const getUsers = ({fromIndex, limit = 50} = {}) => {
  let usersRef = db.ref(couponsPath)

  return new Promise((resolve, reject) => {
    let children = []
    usersRef.limitToFirst(limit).once('value', snp => {
      snp.forEach(csnp => {
        children.push({uid: csnp.key, ...csnp.val()})
      })

      resolve(children)
    })
  })
}

export const makeAdmin = (user) => {
  let adminUserRef = db.ref(`admin/${couponsPath}/${user.uid}`)
  return adminUserRef.set({
    name: user.name,
    email: user.email
  })
}
export const unmakeAdmin = (uid) => {
  let adminUserRef = db.ref(`admin/${couponsPath}/${uid}`)
  return adminUserRef.remove()
}

export const enableUser = (uid) => {
  let ref = db.ref(`${couponsPath}/${uid}`).child('disabled')
  return ref.set(false)
}
export const disableUser = (uid) => {
  let ref = db.ref(`${couponsPath}/${uid}`).child('disabled')
  return ref.set(true)
}

export const setUserProp = (uid, prop, value) => {
  let propRef = getUserRef(uid).child('properties').child(prop)

  return propRef.set(value)
}

export const removeUserProp = (uid, prop, value) => {
  let propRef = getUserRef(uid).child('properties').child(prop)

  return propRef.remove()
}

export const getUserProp = (uid, prop) => {
  let propRef = getUserRef(uid).child('properties').child(prop)

  return propRef.once('value', snp => snp.val())
}

export const getUserProps = (uid, prop, value) => {
  let props = {}
  return new Promise((resolve, reject) => {
    getUserRef(uid).child('properties')
      .once('value', snp => {
        snp.reduce((acc, csnp) => Object.assign(acc, {
          [csnp.key]: csnp.val()
        }), {})

        resolve(props)
      })
  })
  //     let usersRef = db.ref(couponsPath)
  //
  // return new Promise((resolve, reject) => {
  //   let children = []
  //   usersRef.limitToFirst(limit).once('value', snp => {
  //     snp.forEach(csnp => {
  //       children.push({uid: csnp.key, ...csnp.val()})
  //     })
  //
  //     resolve(children)
  //   })
  // })
}

/**
 * Use with onEnter route prop to restrict areas to adminstrative users
 *
 * Usage example:
 *
 * ```jsx
 * import { requireAdmin } from 'osi/user'
 *
 * <Route path='/admin' component={App} onEnter={requireAdmin}>
 * ```
 */
export const requireAdmin = (nextState, replace, done) => {
  getCurrentUser().then(user => {
    isAdmin(user.uid).then(admin => {
      if (!admin) {
        replace('/dashboard') // replace with 403
      }
      done()
    })
  })
}
