// osi/user/index.js
import firebase from '../../../utils/database.react.js'
import { getCurrentUser } from 'osi/auth'

const usersPath = 'users'

const db = firebase.database()

const getUserRef = (uid) => db.ref(`${usersPath}/${uid}`)

getCurrentUser().then(user => {
  // console.log('our user is:', user)
  // console.info('our db is:', db)

  // Check if user is admin
  // isAdmin(user.uid).then(d => {
    // console.log('isAdmin returned', d)
  // })
})

export const createNewUser = ({uid, ...rest}) => {
  const userRef = getUserRef(uid)

  // make sure the `uid` doesn't exist...
  let user = {
    createdOn: firebase.database.ServerValue.TIMESTAMP,
    ...rest
  }

  return userRef.set(user)
    .catch(err => {
      console.error('user add failed:', err)
    })
}

export const isAdmin = uid => {
  // console.info('checking if admin for', uid)
  let adminRef = db.ref(`admin/users/${uid}`)

  return new Promise((resolve, reject) => {
    adminRef.once('value')
    .then(snp => {
      const val = snp.val()
      resolve(val !== null && ((val.gmode && 'GMODE') || true))
    })
    .catch(err => {
      // console.log('admin/users/', uid, 'returned error:', err)
      if (err.code === 'PERMISSION_DENIED') {
        resolve(false)
      } else {
        reject(err)
      }
    })
  })
}

export const getUser = uid => new Promise((resolve, reject) => {
  let prom = db.ref(usersPath).child(uid).once('value')

  prom
    .then(snp => snp.val())
    .then(resolve)
})

export const getUsers = ({fromIndex, limit = 50} = {}) => {
  let usersRef = db.ref(usersPath)

  return new Promise((resolve, reject) => {
    let children = []
    usersRef.limitToFirst(limit).once('value', snp => {
      snp.forEach(csnp => {
        children.push({uid: csnp.key, ...csnp.val()})
      })

      resolve(children)
    })
  })
}

export const makeAdmin = (user) => {
  let adminUserRef = db.ref(`admin/users/${user.uid}`)
  return adminUserRef.set({
    name: user.name,
    email: user.email
  })
}
export const unmakeAdmin = (uid) => {
  let adminUserRef = db.ref(`admin/users/${uid}`)
  return adminUserRef.remove()
}

/**
 * Use with onEnter route prop to restrict areas to adminstrative users
 *
 * Usage example:
 *
 * ```jsx
 * import { requireAdmin } from 'osi/user'
 *
 * <Route path='/admin' component={App} onEnter={requireAdmin}>
 * ```
 */
export const requireAdmin = (nextState, replace, done) => {
  getCurrentUser().then(user => {
    isAdmin(user.uid).then(admin => {
      if (!admin) {
        replace('/dashboard') // replace with 403
      }
      done()
    })
  })
  // isAdmin().then()
  // if (initialized && currentUser && currentUser.uid) done()
  // else if (initialized && (!currentUser || currentUser.uid)) {
  //   replace('/login')
  //   done('Unauthorized Access')
  // } else {
  //   waitToInitialize().then(na => {
  //     if (!currentUser || !currentUser.uid) {
  //       replace('/login')
  //     }
  //     done()
  //   })
  // }
}
