// osi/pages/dashboard/charts/StocksChart.react.js
import React from 'react'
import PropTypes from 'prop-types'
import ReactHighstock from 'react-highcharts/ReactHighstock.src'
import {Link} from 'react-router'
// import Highcharts from 'highcharts';
// import Highlight from 'react-highlight';
import Spinner from '../../../../../components/common/Spinner.react'
import './StocksChart.css'
import mainColors, { darkColors } from 'styleguide/colors'

class StocksChart extends React.Component {
  constructor (props, context) {
    super(props, context)

    this.state = {loading: true}

        // console.log('chartData', this.props.chartData);

    this.timeout = setTimeout(() => {
      this.setState({loading: false})
            // console.log('chartData', this.props.chartData);
    }, 2500)
  }

  componentDidMount () {
        // let chart = this.refs.chart.getChart();
        // chart.series.addPoint({x: 10, y: 12});

        // let chart = this.refs.chart;
        // console.log('refs', chart);
  }

  componentWillUnmount () {
    clearTimeout(this.timeout)
  }

  getChartConfig () {
    const marketData = this.props.chartData.btcAggDelta
    const portfolioData = this.props.chartData.portfolioAggDelta
    var dayDataByDate = this.props.dayDataByDate

        // sort in ascending order
    marketData.sort(function (a, b) {
      let v1 = a[0]
      let v2 = b[0]

      return v1 - v2
    })

    portfolioData.sort(function (a, b) {
      let v1 = a[0]
      let v2 = b[0]

      return v1 - v2
    })

    var seriesOptions = []

    seriesOptions[0] = {
      name: 'BTC ($)',
      data: marketData,
      tooltip: {
        valueDecimals: 2
      },
      marker: {
        enabled: false
      },
      shadow: false,
      lineWidth: 0.5,
      color: mainColors[0][2],
      fillOpacity: 1,
      type: 'areaspline'
    }

    seriesOptions[1] = {
      name: 'Portfolio ($)',
      data: portfolioData,
      tooltip: {
        valueDecimals: 2
      },
      marker: {
        enabled: false
      },
      shadow: false,
      lineWidth: 0.5,
      color: mainColors[0][1],
      fillOpacity: 0.8,
      type: 'areaspline'
    }

    const config = {
      chart: 'area',
      rangeSelector: {
        selected: 1,
        enabled: false
      },
      navigator: {
        enabled: false
      },
      tooltip: {
        formatter: function () {
          let benchmarkValue = this.points[0]
          let portfolioValue = this.points[1]

          let date = new Date(portfolioValue.x)
          let firstDay = new Date(portfolioData[0][0])

                    // prepare balances
          let balances = dayDataByDate[portfolioValue.x].balances
          let balancesStr = '<br/>'
          for (let i = 0; i < balances.length; i++) {
            let balance = balances[i]
            if (balance.balance > 0) {
              balancesStr += '<br/>   *' + balance.symbol + '  ' + Math.round(balance.balance * 100) / 100 + ' (' + Math.round(balance.fiatValue * 100) / 100 + ' ' + balance.fiatCurrency + ')'
            }
          }

          let ret = '<b>Portfolio Value ' + Math.round(dayDataByDate[portfolioValue.x].dayFiatValue * 100) / 100 + ' USD (' + date.toLocaleDateString() + ')</b>'
          ret += '<br/><b>Portfolio Overall Performance: ' + Math.round((portfolioValue.y - 1) * 100) + '% (' + firstDay.toLocaleDateString() + ' - ' + date.toLocaleDateString() + ')'
          ret += '<br/>|<br/>Balances:'
          ret += balancesStr
          ret += '<br/>|<br/><b>Market Overall Performance: ' + Math.round((benchmarkValue.y - 1) * 100) + '% (' + firstDay.toLocaleDateString() + ' - ' + date.toLocaleDateString() + ')</b>'

          return ret
        }
      },
      series: seriesOptions,
      colors: mainColors[3] || darkColors[0],
      xAxis: {
        lineColor: '#fff',
        tickColor: '#fff',
        tickWidth: 0,
        lineWidth: 0,
        labels: {
          style: {
            color: 'rgba(255,255,255,0.7)',
            fontFamily: 'Comfortaa, sans-serif'
          }
        }
      },
      yAxis: {
        gridLineWidth: 0,
        lineColor: '#fff',
        tickColor: '#fff',
        tickLength: 0,
        tickWidth: 0,
        tickPosition: 'outside',
        labels: {
          align: 'left',
          x: 5,
          y: 5,
          style: {
            color: 'rgba(255,255,255,0.7)',
            fontFamily: 'Comfortaa, sans-serif'
          }
        },
        lineWidth: 0
                //  lineColor:'black'
      }
    }
    let theme = {
      colors: ['#2b908f', '#90ee7e', '#f45b5b', '#7798BF', '#aaeeee', '#ff0066', '#eeaaee',
        '#55BF3B', '#DF5353', '#7798BF', '#aaeeee'],
      chart: {
        backgroundColor: {
          linearGradient: { x1: 0, y1: 0, x2: 1, y2: 1 },
          stops: [
            [0, '#f45b5b'],
            [1, '#f45b5b']
          ]
        },
        style: {
          fontFamily: '\'Open Sans\', sans-serif'
        },
        plotBorderColor: '#606063',
        type: 'area'
      },
      title: {
        style: {
          color: '#E0E0E3',
          textTransform: 'uppercase',
          fontSize: '20px'
        }
      },
      subtitle: {
        style: {
          color: '#E0E0E3',
          textTransform: 'uppercase'
        }
      },
      tooltip: {
        backgroundColor: 'rgba(0, 0, 0, 0.85)',
        style: {
          color: '#F0F0F0'
        }
      },
      plotOptions: {
        series: {
          dataLabels: {
            color: '#B0B0B3'
          },
          marker: {
            lineColor: '#333'
          }
        },
        boxplot: {
          fillColor: '#505053'
        },
        candlestick: {
          lineColor: 'white'
        },
        errorbar: {
          color: 'white'
        }
      },
      legend: {
        itemStyle: {
          color: '#E0E0E3'
        },
        itemHoverStyle: {
          color: '#FFF'
        },
        itemHiddenStyle: {
          color: '#606063'
        }
      },
      credits: {
        style: {
          color: '#666'
        }
      },
      labels: {
        style: {
          color: '#707073'
        }
      },

      drilldown: {
        activeAxisLabelStyle: {
          color: '#F0F0F3'
        },
        activeDataLabelStyle: {
          color: '#F0F0F3'
        }
      },

   // scroll charts
      rangeSelector: {
        buttonTheme: {
          fill: '#505053',
          stroke: '#000000',
          style: {
            color: '#CCC'
          },
          states: {
            hover: {
              fill: '#707073',
              stroke: '#000000',
              style: {
                color: 'white'
              }
            },
            select: {
              fill: '#000003',
              stroke: '#000000',
              style: {
                color: 'white'
              }
            }
          }
        },
        inputBoxBorderColor: '#505053',
        inputStyle: {
          backgroundColor: '#333',
          color: 'silver'
        },
        labelStyle: {
          color: 'silver'
        }
      }
    }

    Object.assign(config, theme)
    return config
  }

  render () {
    return (
      <div className='stock-charts'>
        {!this.props.chartData && this.props.exchanges.length && <div><Spinner /><div className='spinner-msg'>This might take a while...</div></div>}
        {!this.props.exchanges.length && <div className='accounts-msg'>Please add <Link to={'/accounts'}>Accounts</Link></div>}
        {this.props.chartData && <ReactHighstock config={this.getChartConfig()} ref='chart' /> }
      </div>
    )
  }
}

StocksChart.propTypes = {
  chartData: PropTypes.object.isRequired
    //, exchanges: PropTypes.array.isRequired
}

export default StocksChart
