// osi/components/portman/sections/presale/index.js
import React, { Component } from 'react'
import { connect } from 'react-redux'
import { browserHistory } from 'react-router'

import { bindActionCreators } from 'redux'

import * as userActions from '../../../../../../actions/user.actions'
import * as portfolioActions from '../../../../../../actions/portfolio.actions'

// import RaisedButton from 'material-ui/RaisedButton'
import Button from 'react-md/lib/Buttons/Button'
import Card from 'react-md/lib/Cards/Card'
import CardTitle from 'react-md/lib/Cards/CardTitle'
import CardText from 'react-md/lib/Cards/CardText'
import Dialog from 'material-ui/Dialog'
import TextField from 'material-ui/TextField'
import ActionAnnouncement from 'material-ui/svg-icons/action/announcement'

import * as Coupons from 'osi/user/coupons'

const RaisedButton = (props) => <Button raised {...props}>{props.children}</Button>

class PresalePage extends Component {
  constructor (props) {
    super(props)

    this.state = {
      createPortfolioOpen: false,
      busy: false
    }
  }

  attemptCoupon (e) {
    const coupon = this.couponValue
    e.preventDefault()

    console.log('do coupon!', coupon)
    console.log('do user!', this.props.user)
    Coupons.isCouponValid(coupon)
      .then(isValid => {
        console.log('is coupon valid?', isValid)
        if (isValid) {
          Coupons.useCoupon(coupon, this.props.user.profile.uid)
            .then(data => {
              console.log('data for coupon:', data)
            })
        }
      })
  }

  addCoupon (e) {
    e.preventDefault()

    Coupons.addCoupon(this.addCouponValue)
      .then(res => {
        console.log('is coupon valid?', res)
      })

  }

  getCouponValue () {
    console.log('getCouponValue')
    Coupons.getCouponValue()
      .then(res => {
        console.log('my coupon value is:', res)
      })
  }

  render () {
    // console.info('portfoliosList:', PortfoliosList)
    return (
      <div className='presale'>
        <h3>Welcome to Presale!</h3>
        <p>Do you have a coupon for Coindash Presale? Use it here!</p>

        <form className='coupons' onSubmit={e => this.attemptCoupon(e)}>
          <legend>Enter Coupon Here</legend>
          <TextField id='coupon-input' floatingLabelText='Enter Coupon' onChange={(e, v) => this.couponValue = v} />
          <RaisedButton onClick={e => this.attemptCoupon(e)} type='submit' label='Attempt Coupon' />
        </form>

        <hr/>
        <hr/>
        <RaisedButton onClick={e => this.getCouponValue(e)} type='submit' label='Show Coupon Value' />

        <hr/>
        <hr/>


        <h4>Admin</h4>

        <form className='coupons' onSubmit={e => this.addCoupon(e)}>
          <legend>Add Coupon Here</legend>
          <TextField id='coupon-input' floatingLabelText='Enter Coupon' onChange={(e, v) => this.addCouponValue = v} />
          <RaisedButton onClick={e => this.addCoupon(e)} type='submit' label='Add Coupon' />
        </form>

      </div>
    )
  }
}

function mapStateToProps (state, ownProps) {
  return {
    user: state.user
  }
}

function mapDispatchToProps (dispatch) {
  return {
    userActions: bindActionCreators(userActions, dispatch)
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(PresalePage)
