// osi/components/portman/parts/calculate-portfolio/index.js
import React, { Component } from 'react'
import AlertDialog from 'osi/components/portman/parts/dialogs/alert-dialog'

class CalculatePortfolio extends Component {
  constructor (props) {
    super(props)

    this.state = {
      alertOpen: false
    }
  }

  startPortfolio () {
    const { portfolio, chartActions, balancesActions, addressList = [] } = this.props
    console.log('starting!', portfolio)
    console.info('addresses:', addressList)

    if (!addressList.length) {
      this.setState({
        alertOpen: true,
        alertTitle: 'No addresses associated',
        alertText: 'Associate addresses with this portfolio to begin calculations'
      })
    } else {
      chartActions.calcPortfolio(portfolio.portfolio.pid, addressList)
      balancesActions.calcBalances(portfolio.portfolio.pid, addressList)
    }
  }

  render () {
    const { processProgress, calculationError, process, button, statusElm } = this.props
    const StartButton = props => button
    const Status = props => statusElm || <div></div>
    return (
      <div className='calculate-wrap'>
        {(!process || !process.started || calculationError || !processProgress || !(processProgress > 0 && processProgress < 100)) && (
          <div onClick={e => this.startPortfolio()}>
            <StartButton />
          </div>
        )}
        {(!calculationError && process && process.started && processProgress && (processProgress > 0 && processProgress < 100)) && (
          <Status />
        )}

        <AlertDialog
          onClose={e => this.setState({alertOpen: false})}
          visible={this.state.alertOpen}
          title={this.state.alertTitle}
          >
          <div>{this.state.alertText}</div>
        </AlertDialog>
      </div>
    )
  }
}

export default CalculatePortfolio


export class CalculatesPortfolio extends Component {
  constructor (props) {
    super(props)

    this.state = {
      alertOpen: false
    }
  }

  startPortfolio () {
    const { portfolio, chartActions, balancesActions, addressList = [] } = this.props
    console.log('starting!', portfolio)
    console.info('addresses:', addressList)

    if (!addressList.length) {
      this.setState({
        alertOpen: true,
        alertTitle: 'No addresses associated',
        alertText: 'Associate addresses with this portfolio to begin calculations'
      })
    } else {
      chartActions.calcPortfolio(portfolio.portfolio.pid, addressList)
      balancesActions.calcBalances(portfolio.portfolio.pid, addressList)
    }
  }

  render () {
    return (
      <div className='calculate-wrap'>
        Hello Calc
      </div>
    )
  }
}
