// osi/components/portman/index.js
import firebase from '../../../../utils/database.react.js'
import { getCurrentUser } from 'osi/auth'
import { usersPath } from 'osi/user'

export const portfoliosPath = 'portfolios'
const db = firebase.database()
const portfoliosRef = db.ref(portfoliosPath)

const tasksPath = 'tasks'

// console.log('firebase:', firebase)
const PUSH_CHARS = '-0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ_abcdefghijklmnopqrstuvwxyz'

export function decodeFirebaseTimestamp (id) {
  id = id.substring(0, 8)
  var timestamp = 0
  for (var i = 0; i < id.length; i++) {
    var c = id.charAt(i)
    timestamp = timestamp * 64 + PUSH_CHARS.indexOf(c)
  }
  return timestamp
}

export const createPortfolio = ({name, uid, description}) => {
  const { TIMESTAMP } = firebase.database.ServerValue
  let newPfid = portfoliosRef.push().key
  let newUserKey = db.ref(`${usersPath}/${uid}/portfolios`).push().key
  let pid = (Math.random() * 10000000000).toString(36).substr(2).split('.').reverse().join('')
  // let newPIDKey = db.ref('pids').child(pid)

  // create a portfolio in `portfolios` and link it to the user at `users/$uid/portfolios`
  let updates = {}
  let portfolio = {
    createdOn: TIMESTAMP,
    pid,
    name,
    owner: uid,
    description
  }
  updates[`${usersPath}/${uid}/portfolios/${newUserKey}`] = newPfid
  updates[`${usersPath}/${uid}/lastUpdate`] = TIMESTAMP
  updates[`${portfoliosPath}/${newPfid}`] = {...portfolio}

  return db.ref().update(updates)
    .then(_ => ({[`pids/${pid}`]: newPfid})) // Create short PID for Portfolio
    .then(r => db.ref().update(r)) // Make the update
    .then(_ => ({ // Return a valid Portfolio object
      pfid: newPfid,
      userKey: newUserKey,
      portfolio
    }))
}

export const getPortfolio = pfid => {
  return portfoliosRef.child(pfid).once('value')
    .then(snp => snp.val())
}

export const getUserPortfolios = (uid, {fromIndex, limit = 50} = {}) => {
  let userPortfoliosRef = db.ref(`${usersPath}/${uid}/portfolios`)

  return new Promise((resolve, reject) => {
    let children = []
    userPortfoliosRef.limitToFirst(limit).once('value', snp => {
      // console.info('snapshot:', snp)
      snp.forEach(csnp => { children.push({key: csnp.key, pfid: csnp.val()}) })
      // console.log('children by now?', children)

      let promises = children.map(
        ({pfid, key}) => getPortfolio(pfid)
          .then(pf => {
            // console.log('result of get portfolio:', pf)
            return {
              pfid,
              portfolio: pf,
              userKey: key
            }
          })
      )

      Promise.all(promises)
        .then(portfolios => {

          resolve(portfolios.filter(pf => pf.portfolio))
        })

      // resolve(children)
    })
  })
}

export const associateAddressWithPortfolio = (portfolio, newAddress) => {
  const { pfid } = portfolio
  console.log('add to database')
  let pfRef = db.ref(portfoliosPath).child(pfid)
  let newAKey = pfRef.child('addresses').push().key
  let newTKey = db.ref(`${tasksPath}/${usersPath}/${portfolio.portfolio.owner}`).child('tasks').push().key

  const updates = {}
  updates[`${portfoliosPath}/${pfid}/addresses/${newAKey}`] = newAddress
  updates[`${tasksPath}/${usersPath}/${portfolio.portfolio.owner}/tasks/${newTKey}`] = {
    createdOn: firebase.database.ServerValue.TIMESTAMP,
    category: 'portfolio',
    type: 'address',
    address: newAddress,
    pfid
  }
  console.log('updates:', updates)
  return db.ref().update(updates)
    .then(_ => {
      console.log('updates successful')
      return true
    })
}
