// osi/app/app-wrap.react.js
import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider'
import getMuiTheme from 'material-ui/styles/getMuiTheme'
import coindashBaseTheme from 'osi/themes/coindashBaseTheme'
import coindashDarkTheme from 'osi/themes/coindashDarkTheme'

const getTheme = (dark) => dark ? coindashDarkTheme : coindashBaseTheme

class AppWrap extends Component {
  render () {
    const Children = props => this.props.children
    return (
      <MuiThemeProvider muiTheme={getMuiTheme(getTheme(this.props.darkTheme))}>
        <div className='app-wrap'><Children /></div>
      </MuiThemeProvider>

    )
  }
  getChildContext () {
    return {
      muiTheme: getMuiTheme(getTheme(this.props.darkTheme))
    }
  }
}
AppWrap.childContextTypes = {
  muiTheme: PropTypes.object
}

function mapStateToProps (state, ownProps) {
  return {
    darkTheme: state.user.darkTheme
  }
}

function mapDispatchToProps (dispatch) {
  return {}
}

export default connect(mapStateToProps, mapDispatchToProps)(AppWrap)
