// osi/app/app.react.js
import React, { Component } from 'react'
import { connect } from 'react-redux'
import 'bootstrap/dist/css/bootstrap.css'
import 'bootstrap/dist/css/bootstrap-theme.css'
import { browserHistory } from 'react-router'

// import Leftnav from './common/leftnav/Leftnav.react'
import Sidenav from '../components/sidenav/sidenav.react'

// import Navbar from './common/navbar/Navbar.react';
import Header from '../components/header/header.react'

import './app.css'

class App extends Component {
  render () {
    return (
      <div className='App'>
        <Header extension={this.props.extension} />
        <div className='main-container'>
          <Sidenav />
          {this.props.children}
        </div>
      </div>
    )
  }

  componentWillUpdate (nextProps, nextState) {
    // Check if user still logged in, navigate to /login if no user ID available
    if (!nextProps.user || !nextProps.user.loggedIn) {
      browserHistory.push('/login')
    }
  }
}

function mapStateToProps (state, ownProps) {
  return {
    extension: state.extension,
    user: state.user
  }
}

function mapDispatchToProps (dispatch) {
  return {

  }
}

export default connect(mapStateToProps, mapDispatchToProps)(App)
