// osi/app/app.react.js
import React, { Component } from 'react'
import { connect } from 'react-redux'
import 'bootstrap/dist/css/bootstrap.css'
import 'bootstrap/dist/css/bootstrap-theme.css'
import { browserHistory } from 'react-router'
import { getCurrentUser } from 'osi/auth'
import { bindActionCreators } from 'redux'

import * as userActions from '../../../actions/user.actions'

import Sidenav from '../components/sidenav/sidenav.react'
import Header from 'osi/components/header2'

import './app.css'

class App extends Component {
  render () {
    return (
      <div className='App'>
        <Header />
        <div className='main-container'>
          <Sidenav />
          {this.props.children}
        </div>
      </div>
    )
  }

  componentWillUpdate (nextProps, nextState) {
    // Check if user still logged in, navigate to /login if no user ID available
    if (!nextProps.user || !nextProps.user.loggedIn) {
      browserHistory.push('/login')
    }
  }
  componentDidMount () {
    // Check if user is logged in but doesn't have user data in state
    getCurrentUser().then(cu => {
      if (!this.props.user || this.props.user.uid !== cu.uid) {
        this.props.userActions.loadUser(cu)
      }
    })
  }
}

function mapStateToProps (state, ownProps) {
  return {
    user: state.user
  }
}

function mapDispatchToProps (dispatch) {
  return {
    userActions: bindActionCreators(userActions, dispatch)
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(App)
