// osi/pages/home/home.react.js
import React, { Component } from 'react'
import { Link } from 'react-router'
import Trianglify from 'react-trianglify'
import FontAwesome from 'react-fontawesome'
import { FormGroup, ControlLabel, FormControl, HelpBlock } from 'react-bootstrap'
import { Page, Card } from 'osi/components'
import { Logo } from 'osi/bits'
import './home.css'
import packageJson from '../../../../../package.json'

function FieldGroup ({ id, label, help, ...props }) {
  return (
    <FormGroup controlId={id}>
      <ControlLabel>{label}</ControlLabel>
      <FormControl {...props} />
      {help && <HelpBlock>{help}</HelpBlock>}
    </FormGroup>
  )
}

class LoginPage extends Component {
  render () {
    let {width, height, seed} = this.state
    let key = `${width}-${height}-${seed}`
    return (
      <Page className='home'>
        <div className='tbg' key={key}>
          <Trianglify width={this.state.width} height={this.state.height} seed={this.state.seed} {...this.state.pattern} />
        </div>
        <Link to='/login' className='brand'>
          <Logo version={packageJson.version} />
        </Link>
      </Page>
    )
  }
  constructor (props) {
    super(props)
    this.state = {
      seed: 'Coindash by Eli Sklar - &copy;2017',
      width: window.innerWidth,
      height: window.innerHeigh,
      pattern: {
        variance: 25,
        cell_size: 50,
        x_colors: ['#fff7fb', '#ece2f0', '#d0d1e6', '#a6bddb', '#67a9cf', '#3690c0', '#02818a', '#016c59', '#014636'],
        y_colors: ['#ffffd9', '#edf8b1', '#c7e9b4', '#7fcdbb', '#41b6c4', '#1d91c0', '#225ea8', '#253494', '#081d58']
      }
    }
  }
  handleChange (event) {
    this.setState({seed: event.target.value})
  }

  componentDidMount () {
    this.updateWindowDimensions()
    window.addEventListener('resize', this.updateWindowDimensions.bind(this))
  }

  componentWillUnmount () {
    window.removeEventListener('resize', this.updateWindowDimensions.bind(this))
  }

  updateWindowDimensions () {
    this.setState({ width: window.innerWidth, height: window.innerHeight })
  }
}

export default LoginPage
